### ISSUE TITLE
Typing: actual type "starlette.testclient.TestClient", expected type "starlite.testing.TestClient"

### ISSUE THREAD

**Issue by peterschutt** [starlite-api/starlite#182]
This _isn't_ related to #174 - I was already ignoring the error before I pulled those changes in.

I cannot see why mypy should complain here:

```python
from collections.abc import Generator
from typing import Any

import pytest
from starlite import CacheConfig, Starlite
from starlite.cache import SimpleCacheBackend
from starlite.testing import TestClient

from app.main import app


@pytest.fixture
def starlite(monkeypatch: Any) -> Starlite:
    monkeypatch.setattr(app, "cache_config", CacheConfig(backend=SimpleCacheBackend()))
    return app


@pytest.fixture
def test_client(starlite: Starlite) -> Generator[TestClient, None, None]:
    with TestClient(app=starlite) as client:
        yield client
```

Error:

```
tests/conftest.py:22: error: Incompatible types in "yield" (actual type "starlette.testclient.TestClient", expected type "starlite.testing.TestClient")  [misc]
```

**Comment by peterschutt** [comment]
Oh, I know.

Starlette test client mustn't have a generic return type for their `__enter__()` ctx mgr method.

**Comment by peterschutt** [comment]
Should we override `TestClient.__enter__()` on our side so that we can override the type?

**Comment by Bobronium** [comment]
> Should we override TestClient.__enter__() on our side so that we can override the type?

Sounds like a plan. 

Let's be smarter than starlette, though:

```py
from typing import TypeVar

TestClientT = TypeVar("TestClientT", bound="TestClient")

...

class TestClient(StarletteTestClient):
    ...

    def __enter__(self: TestClientT) -> TestClientT:
        return super().__enter__()

```

**Comment by peterschutt** [comment]
Way ahead of you;)

**Comment by peterschutt** [comment]
Cleaned up a lot of casts and ignores in our own tests too, which is always nice.

**Comment by peterschutt** [comment]
Closed in #184
